src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WorkerThread.cs

```diff
@@ -0,0 +1,164 @@
+
+using System.Globalization;
+using Internal.LowLevelLinq;
+using Internal.Runtime.Augments;
+
+namespace System.Threading
+{
+    internal static partial class ClrThreadPool
+    {
+        /// <summary>
+        /// The worker thread infastructure for the CLR thread pool.
+        /// </summary>
+        internal static class WorkerThread
+        {
+            /// <summary>
+            /// Semaphore for controlling how many threads are currently working.
+            /// </summary>
+            private static LowLevelLifoSemaphore s_semaphore = new LowLevelLifoSemaphore(0, int.MaxValue);
+
+            private const int TimeoutMs = 20 * 1000;
+            
+            private static void WorkerThreadStart()
+            {
+                // TODO: Event: Worker Thread Start event
+
+                while (true)
+                {
+                    // TODO: Event:  Worker thread wait event
+                    while (s_semaphore.Wait(TimeoutMs))
+                    {
+                        Volatile.Write(ref s_aligned.lastDequeueTime, Environment.TickCount);
+                        if (ThreadPoolWorkQueue.Dispatch())
+                        {
+                            // If we ran out of work, we need to update s_aligned.counts that we are done working for now
+                            // (this is already done for us if we are forced to stop working early in ShouldStopProcessingWorkNow)
+                            ThreadCounts currentCounts = ThreadCounts.VolatileReadCounts(ref s_aligned.counts);
+                            while (true)
+                            {
+                                ThreadCounts newCounts = currentCounts;
+                                newCounts.numProcessingWork--;
+
+                                ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_aligned.counts, newCounts, currentCounts);
+
+                                if (oldCounts == currentCounts)
+                                {
+                                    break;
+                                }
+                                currentCounts = oldCounts;
+                            }
+                        }
+
+                        // Reset thread-local state that we control.
+                        RuntimeThread.CurrentThread.Priority = ThreadPriority.Normal;
+                        CultureInfo.CurrentCulture = CultureInfo.InstalledUICulture;
+                        CultureInfo.CurrentUICulture = CultureInfo.InstalledUICulture;
+
+                        // It's possible that we decided we had no work just before some work came in, 
+                        // but reduced the worker count *after* the work came in.  In this case, we might
+                        // miss the notification of available work.  So we wake up a thread (maybe this one!)
+                        // if there is work to do.
+                        if (ThreadPool.GetQueuedWorkItems().Any())
```

Comment:
This should only be done when Dispatch returns true, after decrementing the worker count. The other case shouldn't have this issue.